cdap-app-fabric/src/main/java/co/cask/cdap/scheduler/ProgramScheduleService.java

```diff
@@ -16,84 +16,227 @@
+   * @return schedules for the given app that match the given predicate
+   * @throws NotFoundException if the application could not be found
+   * @throws UnauthorizedException if the principal is not authorized to access the application
+   * @throws Exception if any other errors occurred while performing the authorization enforcement check
    */
-  public void suspendResumeSchedule(ScheduleId scheduleId, String action) throws Exception {
-    boolean doEnable;
-    if (action.equals("disable") || action.equals("suspend")) {
-      doEnable = false;
-    } else if (action.equals("enable") || action.equals("resume")) {
-      doEnable = true;
-    } else {
-      throw new BadRequestException(
-        "Action for schedules may only be 'enable', 'disable', 'suspend', or 'resume' but is'" + action + "'");
+  public Collection<ProgramScheduleRecord> list(ApplicationId applicationId,
+                                                Predicate<ProgramScheduleRecord> predicate) throws Exception {
+    AuthorizationUtil.ensureAccess(applicationId, authorizationEnforcer, authenticationContext.getPrincipal());
+    return scheduler.listScheduleRecords(applicationId).stream().filter(predicate).collect(Collectors.toList());
+  }
+
+  /**
+   * List the schedules for the given workflow that match the given predicate
+   *
+   * @param workflowId the workflow to get schedules for
+   * @param predicate return schedules that match this predicate
+   * @return schedules for the given program that match the given predicate
+   * @throws UnauthorizedException if the principal is not authorized to access the application
+   * @throws Exception if any other errors occurred while performing the authorization enforcement check
+   */
+  public Collection<ProgramScheduleRecord> list(WorkflowId workflowId,
+                                                Predicate<ProgramScheduleRecord> predicate) throws Exception {
+    AuthorizationUtil.ensureAccess(workflowId, authorizationEnforcer, authenticationContext.getPrincipal());
+    return scheduler.listScheduleRecords(workflowId).stream().filter(predicate).collect(Collectors.toList());
+  }
+
+  /**
+   * Get schedules that are triggered by the given program statuses.
+   *
+   * @param programId the program that is in the trigger
+   * @param statusSet statuses that are involved in the trigger
+   * @return schedules triggered by the given program statuses
+   * @throws UnauthorizedException if the principal is not authorized to access the application
+   * @throws Exception if any other errors occurred while performing the authorization enforcement check
+   */
+  public Collection<ProgramScheduleRecord> findTriggeredBy(ProgramId programId,
+                                                           Set<ProgramStatus> statusSet) throws Exception {
+    AuthorizationUtil.ensureAccess(programId, authorizationEnforcer, authenticationContext.getPrincipal());
```

Comment:
Better to check the authorization on those schedules triggered by this program too.

Reply:
this is a change in behavior that I would rather not do in this refactoring PR. Please open a jira if you think it the current behavior is wrong (seems like it is).

Reply:
We have a JIRA for this already: https://issues.cask.co/browse/CDAP-12650